nasl/nasl_frame_forgery.c

```diff
@@ -173,3 +188,225 @@ nasl_send_arp_request (lex_ctxt *lexic)
+                     const unsigned char *ether_dst_addr)
+{
+  soc_addr_ll->sll_family = AF_PACKET;
+  soc_addr_ll->sll_ifindex = ifindex;
+  soc_addr_ll->sll_halen = ETHER_ADDR_LEN;
+  soc_addr_ll->sll_protocol = htons (ETH_P_ALL);
+  memcpy (soc_addr_ll->sll_addr, ether_dst_addr, ETHER_ADDR_LEN);
+}
+
+/** @brief Prepare message header to be sent with sendmsg().
+ *
+ * @param[out] message The packaged messages to be sent
+ * @param[in] soc_addr_ll The sockaddr_ll structure for capturing
+ * @param[in] payload The payload, a datalink layer frame with payload
+ * @param[in] payload_sz The payload size.
+ */
+static void
+prepare_message (struct msghdr *message, struct sockaddr_ll *soc_addr_ll,
+                 u_char *payload, int payload_sz)
+{
+  struct iovec iov[1];
+  iov[0].iov_base = payload;
+  iov[0].iov_len = payload_sz;
+
+  message->msg_name = soc_addr_ll;
+  message->msg_namelen = sizeof (struct sockaddr_ll);
+  message->msg_iov = iov;
+  message->msg_iovlen = 1;
+  message->msg_control = 0;
+  message->msg_controllen = 0;
+}
+
+/** @brief Forge a datalink layer frame
+ *
+ * @naslparams
+ *
+ * - @n src_haddr     Source MAC address to use.
+ * - @n dst_haddr     Destination MAC address to use.
+ * - @n ether_proto   Ethernet type integer in hex format. Default 0x0800
+ * (ETHER_P_IP)
+ * - @n payload       Payload to be attached to the frame. E.g a forged tcp
+ * datagram.
+ *
+ * - @naslreturn the forged frame.
+ *
+ * @param lexic Lexical context of NASL interpreter.
+ *
+ * @return tree_cell element or null.
+ */
+tree_cell *
+nasl_forge_frame (lex_ctxt *lexic)
+{
+  tree_cell *retc;
+  struct pseudo_frame *frame;
+  u_char *payload = (u_char *) get_str_var_by_name (lexic, "payload");
+  int payload_sz = get_var_size_by_name (lexic, "payload");
+  int frame_sz;
+  char *ether_src_addr = get_str_var_by_name (lexic, "src_haddr");
+  char *ether_dst_addr = get_str_var_by_name (lexic, "dst_haddr");
+  int ether_proto = get_int_var_by_name (lexic, "ether_proto", 0x0800);
+
+  frame = (struct pseudo_frame *) g_malloc0 (sizeof (struct pseudo_frame)
+                                             + payload_sz);
+  memcpy (frame->framehdr.h_dest, ether_dst_addr, ETHER_ADDR_LEN);
+  memcpy (frame->framehdr.h_source, ether_src_addr, ETHER_ADDR_LEN);
+  frame->framehdr.h_proto = htons (ether_proto);
+  frame->payload = payload;
+
+  frame_sz = ETH_HLEN + payload_sz;
+  memcpy ((char *) frame + ETH_HLEN, payload, payload_sz);
+
+  retc = alloc_typed_cell (CONST_DATA);
+  retc->x.str_val = (char *) frame;
+  retc->size = frame_sz;
+  return retc;
+}
+
+/** @brief Send a frame and listen to the answer
+ *
+ * @naslparams
+ *
+ * - @n frame The frame to be sent.
+ * - @n pcap_active     TRUE by default. Otherwise, NASL does not listen
+ * for the answers.
+ * - @n pcap_filter     BPF filter.
+ * - @n pcap_timeout    Capture timeout. 5 by default.
+ *
+ * - @naslreturn Sniffed answer.
```

Comment:
Sniffed answer? Maybe something like response would be better?

Reply:
I have just used _answer_, as it is also used for other modules and to keep the "style". See nasl_packet_forgery.c

Reply:
Do you want me to change it ?

Reply:
I was more concerned about the meaning of sniffed in this context, if it has a deeper meaning. But as it is used in other places too, I would leave it like it is